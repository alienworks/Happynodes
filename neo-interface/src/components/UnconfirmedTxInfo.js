import React, { Component } from 'react'
import { connect } from 'react-refetch'
import axios from 'axios';
import config from './config'

class UnconfirmedTxInfo extends Component {
    constructor(props) {
        super(props);
        this.state = {
            ajaxdone: false
        };
    }

    render() {
        const { unconfirmTxInfo } = this.props

        const assettable = {"0xc56f33fc6ecfcd0c225c4ab356fee59390af8560be0e930faebe74a6daff7c9b":"NEO",
                           "0X602c79718b16e442de58778e148d0b1084e3b2dffd5de6b7b16cee7969282de7":"GAS"}

        if (unconfirmTxInfo.pending) {
            return (
                <br />
            );
        }
        let vins = unconfirmTxInfo.value.data.result.vin


        if (!this.state.ajaxdone) {
            let i
            for (i in vins) {
                console.log("vin", vins[i])
                let vin = vins[i]
                let txid = vin.txid.substring(2)
                let voutidx = vin.vout
                console.log("txid", txid)
                axios.get(`https://neoscan.io/api/main_net/v1/get_transaction/` + txid)
                    .then(res => {
                        console.log(res.data.vouts[voutidx]);
                        let addr = res.data.vouts[voutidx].address_hash
                        let value = res.data.vouts[voutidx].value
                        let asset = res.data.vouts[voutidx].asset

                        unconfirmTxInfo.value.data.result.vin[i].addr = addr
                        unconfirmTxInfo.value.data.result.vin[i].value = value
                        unconfirmTxInfo.value.data.result.vin[i].asset = asset
                        if (i == vins.length - 1) {
                            this.setState({ ajaxdone: true })
                        }
                    })
            }
        }


        console.log("vin", unconfirmTxInfo.value.data.result.vout)

        return (
            <div className="top-block pink">
                <p>blockhash: {unconfirmTxInfo.value.data.result.blockhash}</p>
                <p>blocktime: {unconfirmTxInfo.value.data.result.blocktime}</p>
                <p>type: {unconfirmTxInfo.value.data.result.type}</p>
                {unconfirmTxInfo.value.data.result.vin.map(v => <p key={v.txid} >  vin: {v.addr} {v.value} {v.asset} </p>)}
                {unconfirmTxInfo.value.data.result.vout.map(v => <p key={v.txid} >  vout: {v.address} {v.value} {assettable[v.asset]} </p>)}

                <p>txid: {unconfirmTxInfo.value.data.result.txid}</p>
                <p>size: {unconfirmTxInfo.value.data.result.size} bytes </p>
                <p>gas: {unconfirmTxInfo.value.data.result.gas}</p>
                <p>net_fee: {unconfirmTxInfo.value.data.result.net_fee}</p>
                <p>sys_fee: {unconfirmTxInfo.value.data.result.sys_fee}</p>
                {/* <p>script: {unconfirmTxInfo.value.data.result.script}</p> */}

            </div>
        );
    }
}

export default connect((props) => ({
    unconfirmTxInfo: {
        method: 'POST',
        url: config.api_url.replace('/redis', '').concat(`/unconfirmed/tx`),
        body: JSON.stringify({ 'tx': props.tx, 'addressid': props.addressid })
    }
}))(UnconfirmedTxInfo)