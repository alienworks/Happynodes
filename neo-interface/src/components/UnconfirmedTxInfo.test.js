import React from 'react';
import ReactDOM from 'react-dom';
import UnconfirmedTxInfo from './UnconfirmedTxInfo';
import renderer from 'react-test-renderer';

it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<UnconfirmedTxInfo />, div);
    ReactDOM.unmountComponentAtNode(div);
});

var data = {"data":{"jsonrpc":"2.0","id":1,"result":{"txid":"0x0f2db24bf1fd93c4da75d390fc5733fdc5a8e9c04a966a30bc9195a4448519e0","size":202,"type":"ContractTransaction","version":0,"attributes":[],"vin":[{"txid":"0x550ac998ce924810fca35739dd609a2cbd295d490d370f20bbd4b117a8f7550e","vout":1}],"vout":[{"n":0,"asset":"0xc56f33fc6ecfcd0c225c4ab356fee59390af8560be0e930faebe74a6daff7c9b","value":"1","address":"AWH581PR16ARnu7sPWdx8Gzvpb7hWXcj64"}],"sys_fee":"0","net_fee":"0","scripts":[{"invocation":"407f697652e095575b5b63ff4fbf86bb7b6fc79ce150e17e344ee8a0213500a46f912c8b5078c34a13e4eaa985e1a9c99786b82e57a6086699963af80c94655c43","verification":"2103279039fa598ab5ca11e47d0f1b957666eaa989cb729c426344e66cb60910136dac"}]}}}


describe("test against valid snapshot", ()=> {
    it("redner as expected", ()=> {
        const tree = renderer.create(<UnconfirmedTxInfo unconfirmTxInfo={{fulfilled:true,value:data}} />).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
  