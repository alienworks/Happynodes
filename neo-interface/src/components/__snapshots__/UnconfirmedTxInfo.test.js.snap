// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test against valid snapshot redner as expected 1`] = `
<div
  className="jumbotron"
>
  <h2>
    Transaction Info for 
    0x0f2db24bf1fd93c4da75d390fc5733fdc5a8e9c04a966a30bc9195a4448519e0
  </h2>
  <hr
    style={
      Object {
        "borderColor": "#78cadd",
        "borderWidth": "3px",
      }
    }
  />
  <table
    className="table"
  >
    <thead>
      <tr>
        <th
          className="tx_td"
        >
          VINs Transaction List (i.e. Preceding linked Transactions)
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          className="tx_td"
          style={
            Object {
              "padding": "1rem",
            }
          }
        >
          <a
            href="./550ac998ce924810fca35739dd609a2cbd295d490d370f20bbd4b117a8f7550e"
          >
            0x
            550ac998ce924810fca35739dd609a2cbd295d490d370f20bbd4b117a8f7550e
          </a>
        </td>
      </tr>
    </tbody>
  </table>
  <pre>
    {
        "data": {
            "jsonrpc": "2.0",
            "id": 1,
            "result": {
                "txid": "0x0f2db24bf1fd93c4da75d390fc5733fdc5a8e9c04a966a30bc9195a4448519e0",
                "size": 202,
                "type": "ContractTransaction",
                "version": 0,
                "attributes": [],
                "vin": [
                    {
                        "txid": "0x550ac998ce924810fca35739dd609a2cbd295d490d370f20bbd4b117a8f7550e",
                        "vout": 1
                    }
                ],
                "vout": [
                    {
                        "n": 0,
                        "asset": "0xc56f33fc6ecfcd0c225c4ab356fee59390af8560be0e930faebe74a6daff7c9b",
                        "value": "1",
                        "address": "AWH581PR16ARnu7sPWdx8Gzvpb7hWXcj64"
                    }
                ],
                "sys_fee": "0",
                "net_fee": "0",
                "scripts": [
                    {
                        "invocation": "407f697652e095575b5b63ff4fbf86bb7b6fc79ce150e17e344ee8a0213500a46f912c8b5078c34a13e4eaa985e1a9c99786b82e57a6086699963af80c94655c43",
                        "verification": "2103279039fa598ab5ca11e47d0f1b957666eaa989cb729c426344e66cb60910136dac"
                    }
                ]
            }
        }
    }
  </pre>
</div>
`;
